toolchain/check/handle_class.cpp

```diff
@@ -695,7 +695,36 @@ auto HandleParseNode(Context& context, Parse::ClassDefinitionId node_id)
   auto& class_info = context.classes().Get(class_id);
   class_info.complete_type_witness_id = complete_type_witness_id;
 
-  context.inst_block_stack().Pop();
+  int32_t field_index = class_info.is_dynamic;
+
+  // This is pretty subtle code, not something we'd want to endorse in general -
```

Comment:
I think if we go in this direction, it might make sense to build a list of `InstId`s of the `FieldDecl`s as we process the class, instead of building a list of `StructTypeField`s. Then in `CheckCompleteClassType` we can build the `StructTypeField`s for the `FieldDecl`s and assign indexes to them in the same pass.

Reply:
To lower the cost/avoid iterating through all the instructions in the class? (the usual time/space (storage of this separate list) tradeoff)

Sure, happy to implement that if/when the decision is made in this direction.

Reply:
I was imagining that the list would replace the current `struct_type_fields` tracking, so we could do it without extra cost. While avoiding the iteration through the class body is a nice benefit, I was actually mostly thinking of doing this to make it clearer that the `ElementIndex`es correspond to the `StructTypeField` indexes.

Reply:
Ah, OK, I'm with you. I can look into that. That way it'd be clearer that the `FieldDecl`s wouldn't be "ready" until the end of the class (they wouldn't be in a bundle of instructions until they were assigned their indexes correctly) and the struct fields wouldn't even exist until then, so the tombstoned indexes would be useless anyway - there'd be nothing to index into.

Reply:
Tried this in 28f1ab6, which presented a few questions
* Should this new `field_decls_stack` contain the actual FieldDecls, delaying when they are constructed into instructions and added to the class's inst block until the end of the class checking? For now I've not done that - the FieldDecl instructions are created, added to the inst block, and their InstIds are stored in `field_decls_stack`. This does mean the FieldDecl's indexes do need to be updated (with `insts().Set(field_decl_id, field_decl);`)
* How should the BaseDecl be handled? The InstId could be added to the `field_decls_stack` along with the field decls InstIds, but for now I'm not doing that and special casing it in `CheckCompleteClassType` to add the base subobject to the object representation field list.
* Getting the type for the object representation fields from the FieldDecls was a bit tricky - the FieldDecl's type is an `UnboundElementType`, so I decomposed that and retrieved its `element_type_id` - this isn't possible if the FieldDecl's type is an error, in which case I made the object representation's field an Error type (but this is different from what happens without this patch, where the underlying type is available - that could be stored as extra data in the `field_decls_stack` if we wanted to).
  * Is this assuming that a `FieldDecl`s type either an error or `UnboundElementType` reasonable? Or too brittle? Is there a better approach?

(also, more broadly: Any thoughts on how to figure out which of the 3 directions to move forward with?)

Reply:
From today's toolchain meeting, sounds like I'm on roughly the right track with these issues, but should make sure that the FieldDecls are expected to be only UnboundElementType or Errors - if they're anything else we should assert here, in case other options come up later & we need to revisit this code. I'll double check/work on ensuring that's the case.

Reply:
Modified the code to ensure it'll check-fail if the `FieldDecl`'s `type_id` is anything other than `Error` or `UnboundElementType` in https://github.com/carbon-language/carbon-lang/pull/4514/commits/9038d0c52924e9e3c79509e4eb763f4f16345d3d

So, I think that's covered everything I'm aware of/this patch is ready for more feedback.